crates/autopilot/src/decoded_settlement.rs

```diff
@@ -0,0 +1,487 @@
+    model::order::OrderKind,
+    num::BigRational,
+    number_conversions::big_rational_to_u256,
+    shared::{conversions::U256Ext, external_prices::ExternalPrices},
+    web3::ethabi::{Function, Token},
+};
+
+// Original type for input of `GPv2Settlement.settle` function.
+type DecodedSettlementTokenized = (
+    Vec<Address>,
+    Vec<U256>,
+    Vec<(
+        U256,            // sellTokenIndex
+        U256,            // buyTokenIndex
+        Address,         // receiver
+        U256,            // sellAmount
+        U256,            // buyAmount
+        u32,             // validTo
+        Bytes<[u8; 32]>, // appData
+        U256,            // feeAmount
+        U256,            // flags
+        U256,            // executedAmount
+        Bytes<Vec<u8>>,  // signature
+    )>,
+    [Vec<(Address, U256, Bytes<Vec<u8>>)>; 3],
+);
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct DecodedSettlement {
+    // TODO check if `EncodedSettlement` can be reused
+    pub tokens: Vec<Address>,
+    pub clearing_prices: Vec<U256>,
+    pub trades: Vec<DecodedTrade>,
+    pub interactions: [Vec<DecodedInteraction>; 3],
+}
+
+#[derive(Debug)]
+pub struct DecodedTrade {
+    pub sell_token_index: U256,
+    pub buy_token_index: U256,
+    pub receiver: Address,
+    pub sell_amount: U256,
+    pub buy_amount: U256,
+    pub valid_to: u32,
+    pub app_data: Bytes<[u8; 32]>,
+    pub fee_amount: U256,
+    pub flags: U256,
+    pub executed_amount: U256,
+    pub signature: Bytes<Vec<u8>>,
+}
+
+#[derive(Debug)]
+pub struct DecodedInteraction {
+    pub target: Address,
+    pub value: U256,
+    pub call_data: Bytes<Vec<u8>>,
+}
+
+impl From<(Address, U256, Bytes<Vec<u8>>)> for DecodedInteraction {
+    fn from((target, value, call_data): (Address, U256, Bytes<Vec<u8>>)) -> Self {
+        Self {
+            target,
+            value,
+            call_data,
+        }
+    }
+}
+
+impl From<DecodedSettlementTokenized> for DecodedSettlement {
+    fn from((tokens, clearing_prices, trades, interactions): DecodedSettlementTokenized) -> Self {
+        DecodedSettlement {
+            tokens,
+            clearing_prices,
+            trades: trades
+                .into_iter()
+                .map(|trade| DecodedTrade {
+                    sell_token_index: trade.0,
+                    buy_token_index: trade.1,
+                    receiver: trade.2,
+                    sell_amount: trade.3,
+                    buy_amount: trade.4,
+                    valid_to: trade.5,
+                    app_data: trade.6,
+                    fee_amount: trade.7,
+                    flags: trade.8,
+                    executed_amount: trade.9,
+                    signature: trade.10,
+                })
+                .collect(),
+            interactions: interactions.map(|inner| inner.into_iter().map(Into::into).collect()),
+        }
+    }
+}
+
+pub struct FeeConfiguration {
+    pub fee_objective_scaling_factor: f64,
```

Comment:
When does this value get populated? I assume it just takes whatever value the `autopilot` was started with. That would mean that we can encounter inconsistencies when:
1. some solver wins right to execute
2. solver tries to bring settlement on-chain
3. we restart the `autopilot` with a different fee scaling value
4. we notice the settlement got mined and we index it with the new fee scaling value opposed to the old one

This is error case would happen almost (or actually) never since we rarely change the fee scaling factor, but I still wanted to bring it up to document it.

Reply:
Interesting case. But it feels like it is not a big deal since it just affects the EXACT TIMESTAMP from which we start to calculate the `fee` using the updated parameter.